### DISCUSSION TITLE
genzor.staticweb - lets talk design and requirements

### DISCUSSION THREAD

**Discussion by egil** [egil/genzor#7] [Ideas]
@tesar-tech [already prototyped](https://github.com/tesar-tech/GenzorStaticSiteGenerator) on this. Lets go over the details and see what is missing to make this a nice experience.

First thing that comes to mind is to utilizing the `@page` directive to indicate that the following component is a "text file" instead of the explicit `<TextFile>` components.

**Comment by tesar-tech** [comment]
Going through my experiment I hit some questions and issues:

- How to add static files like css and imgs.
  - I solved this while writing this comment. Not the hard one.. (see my repo)
- How to insert script inside razor components?
  - Inserting the script inside a .razor results in: `Script tags should not be placed inside components because they cannot be updated dynamically. To fix this, move the script tag to the 'index.html' ` 
  - So we are looking for a way how to use index.html, that is not razor component..(?)
- More elegant way how to pass metadata of .md files (like `title` ,`dateTime`, etc... )
  - YAML is the answer: https://github.com/aaubry/YamlDotNet

**Reply by egil** [reply]
1. static files can probably be handled in multiple ways. What was your solution?
2. If I remember correctly, that warning about script tags comes from the Blazor compiler, which means it runs before Genzor gets a chance to do anything about it. But we can always get around that by creating custom components that creates the render tree manually, I think.
3. Agree that a YAML preamble in .md files seems to be a common approach. We could also look to DocFx for inspiration. It also has a bunch of features for creating stable cross references between files.
4. Another thing that could be useful would be a live/watch mode, where we monitor source files (.md) for changes and automatically re-render.
5. Is there a need to support asynchronous operations while generating the site?

Also, I have not looked much at static site generators for a while, so not up to date on the features those should have nowadays. Can you list the overall requirements you are thinking of? I want to make sure the the core genzor lib can support those.

**Reply by tesar-tech** [reply]
1. Just adding the folder with resources to project and naming top generated folder with same name... It will create new files inside the folder. 

Overall I kind of feel a lot of these issues are already solved in https://github.com/statiqdev . It would be unproductive to reinvent the wheel. We can be inspired by that or maybe create some genzor module for statiq or I don't know.. Unfortunately, my knowledge is not deep enough to assess the best solution...

I was astonished how easily the demo page was done with genzor, but I currently don't have static web scenario to be done.

**Comment by egil** [comment]
Instead of writing content in markdown files, I've been thinking that maybe content should be written in markdown INSIDE razor files. Then when the genzor generator runs, it first renders the razor files to text, and then runs a markdown parser/generator on that text.

What is the advantage?
You can mix markdown when Blazor/razor syntax to build the resulting content of a page. 

That means things like cross page references can be strongly typed, e.g. want to reference another page, add a `<Page Ref="OtherPageComponentType" />`. 

Other things like front matter, which would normally be in yaml at the top of the content file can be strongly typed, e.g. as simple C# properties in a code block.

```razor
@page url
@code
{
  public string Title => "My Page";
  public string MetaDescriptiob => ...
}

# Heading 1 #
My content
```

**Reply by egil** [reply]
That might also make it much easier to get hot reload to work out of the box.

**Reply by tesar-tech** [reply]
I see. The statiq also supports `.md` and `.cshtml` files. The thing about markdown and yaml is, that it is kind of standard. Supporting only .razor files would make migrating from/to other tools bit more complicated. But I can imagine how writing only `.razor` would be much easier.

**Reply by egil** [reply]
I think we can support both. If the file is written i .razor, it will first be "rendered", and then the result from it will be transformed using the markdown compiler, who can also process a YAML frontmatter if needed.

About converting, I am not to worried. The docs site for bUnit, which is built using DocFX is pretty comprehensive, but I think for the most part it would be a matter of renaming the .md files to .razor and replacing the yaml frontmatter with the appropraite `@code { ... }` block.

But I do see your point about making it not necessary to do so.

**Reply by egil** [reply]
Btw. didn't know about Statiq. It looks good. It makes me wonder creating static websites with genzor makes sense then. Or is Genzor with the Blazor Component model approach unique enough to warrant the effort?
